module/zfs/dsl_deadlist.c

```diff
@@ -572,3 +771,118 @@ dsl_deadlist_move_bpobj(dsl_deadlist_t *dl, bpobj_t *bpo, uint64_t mintxg,
 	}
 	mutex_exit(&dl->dl_lock);
 }
+
+typedef struct livelist_entry {
+	const blkptr_t *le_bp;
+	avl_node_t le_node;
+} livelist_entry_t;
+
+static int
+livelist_compare(const void *larg, const void *rarg)
+{
+	const blkptr_t *l = ((livelist_entry_t *)larg)->le_bp;
+	const blkptr_t *r = ((livelist_entry_t *)rarg)->le_bp;
+
+	/* Sort them according to dva[0] */
+	uint64_t l_dva0_vdev, r_dva0_vdev;
+	l_dva0_vdev = DVA_GET_VDEV(&l->blk_dva[0]);
+	r_dva0_vdev = DVA_GET_VDEV(&r->blk_dva[0]);
+	if (l_dva0_vdev < r_dva0_vdev)
+		return (-1);
+	else if (l_dva0_vdev > r_dva0_vdev)
+		return (+1);
```

Comment:
Sounds like a good idea, though I guess this comparison is a little more complex different since it's based on two factors (and we have that assertion that entries at the same location have the same birth.) 

Here's what I changed it to - let me know if you have anything more efficient in mind. 

```
793         if (l_dva0_vdev != r_dva0_vdev)
794                 return (AVL_CMP(l_dva0_vdev, r_dva0_vdev));   
            ...
801         if (l_dva0_offset == r_dva0_offset)
802                 ASSERT3U(l->blk_birth, ==, r->blk_birth);
804         return (AVL_CMP(l_dva0_offset, r_dva0_offset));
```

Reply:
That's nice.  My only suggestion would be replace the second conditional for the ASSERT with an IMPLY.  Though, it likely doesn't matter since I'd expect the compiler to entirely optimize out that code out for non-debug builds.  So I'm happy with it either way.  Your version does have the advantage that it will log the bad blk_birth values.

```diff
-         if (l_dva0_offset == r_dva0_offset)
-                 ASSERT3U(l->blk_birth, ==, r->blk_birth);
+         IMPLY(l_dva0_offset == r_dva0_offset, l->blk_birth == r->blk_birth);
```